torch/_dynamo/output_graph.py

```diff
@@ -2229,6 +2198,274 @@ def maybe_lift_tracked_freevar_to_input(self, arg):
+            if need_bind(s):
+                log.debug(
+                    "_lazy_bind_unbacked_symbols_in_tensors %s for %s.stride()[%s] at debug_level %s",
+                    s,
+                    t_proxy,
+                    i,
+                    tracer.debug_level,
+                )
+                lazy_proxy = SubgraphTracer.LazyProxy(
+                    _proxy_with_example_value,
+                    s,
+                    "call_function",
+                    torch.ops.aten.sym_stride.int,
+                    (t_proxy, i),
+                    {},
+                    type_expr=type(s),
+                )
+                assert isinstance(s, torch.SymInt)  # to make linter happy
+                tracer.bound_symbols[s.node.expr] = lazy_proxy
+
+    # NOTE: [Auto lift basic free symbols when create_graph_input]
+    # Whenever we call create_graph_input, we try to also lift the basic symbols in example values
+    # as graph input.
+    # This applies to both top-level graph and subgraphs in higher order ops.
+    # This has several cases:
+    #  1. When create_graph_input for a tensor that has symbolic shapes,
+    #     we look for basic symbols in its size and stride, we check if the symbol is bound
+    #     in current graph (i.e. bound_symbols), it it's not bound, we'll create a placeholder
+    #     for it then recursively check its parent, creates ph if not bound.
+    #     Every tracer maintains a mapping (i.e. lifted_freevars)
+    #     that maps from parent proxy to proxy in current tracer for the symbol.
+    #  2. When create_graph_input for a tensor with unbacked symbolic shapes,
+    #     Backed symbols all come from inputs's symbolic shape. But unbacked symbols
+    #     can be created while tracing. So we use lazy_bind_unbacked_symbols will intercept
+    #     at wrap_fx_proxy, and try to bind the unbacked symbols immediately after they're
+    #     created.
+    #  3. subgraph will also lifted basic symbols in compound exprs.
+    #     For example, if an input to subgraph takes size [s1+s2//8], we'll look for the
+    #     the free symbols in the sizes and lift as inputs similar to 1 in _lift_symbols_in_symint)
+    # Do not call this function direclty. This is used in create_graph_input for
+    # automatically lifting the free basic symbols in example_value.
+    def _lift_basic_symbols(self, example_value, src: Optional[Source]):
+        # See note [Export inputs must be explicitly passed in]
+        is_strict_export = self.is_export
+        is_non_strict_export = torch.compiler.is_compiling()
+        if is_strict_export or is_non_strict_export:
+            return
+
+        if isinstance(example_value, torch.Tensor):
+            return self._lift_symbols_in_tensor(example_value, src)
+
+    # See Note [Auto lift basic free symbols when create_graph_input]
+    def _lift_symbols_in_tensor(self, t: torch.Tensor, src: Optional[Source]):
+        for i, s in enumerate(t.size()):
```

Comment:
Is this new code or did you move it from somewhere?

Reply:
Yeah, this is moved from add_symbol_bindings and did a bit refactoring